File: HashTables/Lab/C-Code/hashDictionary.c
/**
 * hashDictionary.c
 *
 * Implements a dictionary using an external hash table
 *
 * Created by Sally Goldin 1 April 2013 as a sample solution for
 *   CPE113 Lab 10
 *   
 *   Modified 27 March 2017 to change bucket count to prime number
 * Adapted for AiCE WARP 15 July 2022
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "chainedHashTable.h"

#define BUCKET_COUNT 53

/* Structure to store one word with its definition */
/* This will be the data we store in the hash table */
typedef struct
{
    char word[64];      /* word we want to store */
    char definition[1024];  /* definition */
} DICT_ENTRY_T;
 

/* Read word file and construct the dictionary.
 * Arguments
 *  pFile      -  file pointer for reading, assumed open and checked.
 * Returns 1 if successful, 0 for allocation failure or read error.
 */
int buildDictionary(FILE* pFile)
{
   char input[512];      /* buffer to store word and definition from file */
   char* pDelim = NULL;  /* used to find position of '|' */
   DICT_ENTRY_T * pEntry = NULL;
   int returnVal = 1;
   int wordCount = 0;
   while ((fgets(input,sizeof(input),pFile) != NULL) &&
           (returnVal == 1))  
   {
      int len = strlen(input);
      if (input[len-1] == '\n')
      {
         input[len-1] = '\0';    /* get rid of newline */
      }
      pDelim = strpbrk(input,"|");  /* find the vertical bar */
      if (pDelim != NULL)           /* if no delim, just skip the line */
      {
         *pDelim = '\0';              /* replace ':' with 0 */
         pEntry = (DICT_ENTRY_T*) calloc(1,sizeof(DICT_ENTRY_T));
         if (pEntry == NULL) /* allocation error */
         {
            returnVal = 0;
            break; /* this is not a recoverable error so exit */
         }
         *pDelim = '\0'; /* replace delimiter with a terminator */
         strcpy(pEntry->word,input);  /* copy the word */
         strcpy(pEntry->definition,(pDelim+1)); /* copy the definition */
         /* Put the entry into the dictionary */
         returnVal = hashTableInsert(pEntry->word,pEntry);
         if (!returnVal)  /* success */
         {
            printf("Error inserting %s into hash table\n", pEntry->word);
            break;
         }
         wordCount++;
      }  /* if no delimiter, just skip the line */
   }
   printf("Stored %d words in the dictionary\n",wordCount);
   return returnVal;
}

/* print the definitions for a word, or print 'Not Found'
 * if not in the dictionary. 
 * Arguments 
 *   word          -   Word to look up
 */
void printDefinition(char* word)
{
   DICT_ENTRY_T* pWord = NULL;
   printf("\n");
   pWord = (DICT_ENTRY_T*) hashTableLookup(word);
   if (pWord == NULL)
   {
      printf("Word '%s' not found in the dictionary\n",word);
   }
   else
   {
      int i;
      printf("Word: %s \n",pWord->word);
      printf("Definition: %s\n",pWord->definition); 
   }
   printf("\n");
}  


int main(int argc, char* argv[])
{

   FILE * pFp = NULL;           /* used to read input file */
   int status = 0;              /* for function return values */
   pFp = fopen("wordlist.txt","r");
   if (pFp == NULL)
   {
      printf("Error - cannot open file 'wordlist.txt'\n");
      exit(1);
   } 
   /* initialize the hash table */
   status = hashTableInit(BUCKET_COUNT);
   if (!status)
   {
      printf("Error initializing hash table\n");
      exit(1);
   }
   status = buildDictionary(pFp);
   fclose(pFp);
   if (!status)
   {
      printf("Allocation error building dictionary\n");
      exit(1);
   } 
   /* now loop, allowing lookup or addition of new words */
   while (1)
   {
      int option = 0;
      char input[64];
      char word[64];
      printf("\nAvailable operations: \n");
      printf("\t1 - Look up a word in the dictionary\n");
      printf("\t2 - Add a word to the dictionary\n");
      printf("\t3 - Add an additional definition to an existing word\n");
      printf("\t4 - Exit the program\n");
      printf("Option? ");
      fgets(input,sizeof(input),stdin);
      sscanf(input,"%d",&option);
      if (option == 1) 
      {
         printf("Find definition for what word ('EXIT' to exit)? ");
         fgets(input,sizeof(input),stdin);
         memset(word,0,sizeof(word));
         sscanf(input,"%s",word);
         printDefinition(word);
      }    
      else if (option == 2)
      {
         printf("FOR IMPLEMENTATION BY THE STUDENT!\n");
         // STUDENT TASK ONE
         // Add code here to ask for word and definition, then
         // add to the dictionary
      }
      else if (option == 3)
      {
         printf("FOR IMPLEMENTATION BY THE STUDENT!\n");
         // STUDENT TASK TWO
         // Add code here to add an additional definition to an existing
         // word. NOTE you may need to modify the structure of
         // DICT_ENTRY_T in order to complete this task
      }
      else if (option == 4)
      {
         break;
      }
      else 
      {
         printf("Please enter a value between 1 and 4!\n");
      }
   }
   /* free the hash table before we exit */
   hashTableFree();
}

File: HashTables/Lab/Java-Code/HashDictionary.java
/**
 *  HashDictionary.java
 * 
 *  Main class for laboratory exercise using a hash table to build a dictionary.
 * 
 *  Created by Sally Goldin, 15 July 2022, for AiCE Warp
 */ 
import java.util.Hashtable;
import java.util.Scanner;

public class HashDictionary
{
   /** class instance for reading the input file */
   private static TextFileReader reader = new TextFileReader();

   /** object to use for getting terminal input */
   private static Scanner scan = new Scanner(System.in);

   /** Hashtable to serve as our dictionary */
   private static Hashtable<String,DictionaryEntry> dictionary = new Hashtable<String,DictionaryEntry>();

   /** 
    * Ask the user a question and get a string as a reply
    * @param  prompt --  Question or message
    * @return  answer from the user
    */
   public static String promptForString(String question)
   {
      System.out.print(question +" ");
      String input = scan.nextLine();
      return input;
   }

   /** 
    * Read word file and construct the dictionary.
    * @param  filename    Name of file to read
    * Returns 1 if successful, 0 for I/O error.
    */
   public static int buildDictionary(String filename)
   {
      DictionaryEntry entry = null;
      String line;
      int returnVal = 1;
      int wordCount = 0;
      if (!reader.open(filename)) 
      {
         System.out.println("Error opening word file '" + filename + "'\n");
         return 0;
      }
      // otherwise read each line and process it
      while (((line = reader.getNextLine()) != null) &&
           (returnVal == 1))
      {
         int delim = line.indexOf("|");
         if (delim >= 0)
         {
            String word = line.substring(0,delim);
            String definition = line.substring(delim+1);
            entry = new DictionaryEntry(word,definition);
            dictionary.put(word,entry);
            wordCount++;
         }
      }  /* if no delimiter, just skip the line */
      reader.close();
      System.out.println("Stored " + wordCount + " words in the dictionary\n");
      return returnVal;
   }

   /** 
    * Look up a word and 
    * print its definition(s) for a word, or print 'Not Found'
    * if not in the dictionary. 
    * @param  word          -   Word to look up
    */
   private static void printDefinition(String word)
   {
      DictionaryEntry entry = null;
      entry = dictionary.get(word);
      if (entry == null)
      {
         System.out.println("Word '" + word + "' not found in the dictionary\n");
      }
      else
      {
         entry.printInfo();
      }
   }  

   /**
    * Main function presents menu and controls logic
    */
   public static void main(String args[])
   {
      int status = buildDictionary("wordlist.txt");
      if (status == 0)
      {
         System.out.println("Cannot read file\n");
         System.exit(1);
      } 
      /* now loop, allowing lookup or addition of new words */
      while (true)
      {
         int option = 0;
         String input;
         String word;
         System.out.println("\nAvailable operations: ");
         System.out.println("\t1 - Look up a word in the dictionary");
         System.out.println("\t2 - Add a word to the dictionary");
         System.out.println("\t3 - Add an additional definition to an existing word");
         System.out.println("\t4 - Exit the program");
         input = promptForString("Option? ");
         try
         {
            option = Integer.parseInt(input);
         }
         catch (NumberFormatException nfe)
         {
            System.out.println("Please enter a valid integer between 1 and 4\n");
            continue;
         }
         if (option == 1) 
         {
            word = promptForString("Find definition for what word? ");
            printDefinition(word);
         }
         else if (option == 2)
         {
            System.out.println("FOR IMPLEMENTATION BY THE STUDENT!\n");
            // STUDENT TASK ONE
            // Add code here to ask for word and definition, then
            // add to the dictionary
         }
         else if (option == 3)
         {
            System.out.println("FOR IMPLEMENTATION BY THE STUDENT!\n");
            // STUDENT TASK TWO
            // Add code here to add an additional definition to an existing
            // word. NOTE you may need to modify the structure of
            // DICT_ENTRY_T in order to complete this task
         }
         else if (option == 4)
         {
            break;
         }
         else 
         {
            System.out.println("Please enter a value between 1 and 4!\n");
         }
      }
   }

}